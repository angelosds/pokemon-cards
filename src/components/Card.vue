<template>
  <article class="card">
    <div class="card__container" v-if="imageUrl">
      <img class="card__image" loading="lazy" :alt="name" :src="imageUrl" />
    </div>

    <div class="card__description">
      <h1 class="card__name">{{ name }}</h1>
      <span class="card__number">{{ number }}</span>

      <ul class="card__types" v-if="types">
        <li class="card__item" v-for="type in types" :key="type">
          <badge class="card__type" :color="type">{{
            $t(`cardTypes.${type}`)
          }}</badge>
        </li>
      </ul>

      <primary-button class="card__button" @on-click="onClick">{{
        buttonText
      }}</primary-button>
    </div>
  </article>
</template>

<script lang="ts">
import { Component, Emit, Prop, Vue } from "vue-property-decorator";

import Badge from "@/components/Badge.vue";
import PrimaryButton from "@/components/PrimaryButton.vue";

import PokemonType from "@/enums/pokemonType";

@Component({
  components: {
    Badge,
    PrimaryButton
  }
})
export default class Card extends Vue {
  @Prop() readonly buttonText: string;
  @Prop() readonly imageUrl?: string;
  @Prop() readonly name: string;
  @Prop() readonly number?: number;
  @Prop() readonly types?: Array<PokemonType>;

  @Emit()
  onClick(event: Event) {
    return event;
  }
}
</script>

<style lang="scss" scoped>
$card-padding: 16px;

.card {
  border-radius: 16px;
  box-shadow: 0 0 24px rgba(0, 0, 0, 0.1);
  display: flex;
  flex-direction: column;
  overflow: hidden;

  &__button {
    grid-area: button;
  }

  &__container {
    display: flex;
    justify-content: center;
    padding: $card-padding;
  }

  &__description {
    background-color: $secondary-color;
    display: grid;
    grid-gap: 8px;
    grid-template-columns: 3fr 1fr;
    grid-template-areas:
      "name number"
      "types types"
      "button button";
    padding: $card-padding;
  }

  &__image {
    border-radius: 8px;
    height: 330px;
    object-fit: contain;
    width: 240px;
  }

  &__item {
    display: flex;

    &:not(:first-child) {
      margin-left: 4px;
    }

    &:not(:last-child) {
      margin-right: 4px;
    }
  }

  &__name {
    font-size: 2rem;
    grid-area: name;
    margin: 0;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
  }

  &__number {
    font-size: 2rem;
    font-weight: 700;
    grid-area: number;
    justify-self: end;
  }

  &__types {
    display: flex;
    grid-area: types;
    justify-content: center;
    margin: 0;
    padding: 0;
  }
}
</style>
