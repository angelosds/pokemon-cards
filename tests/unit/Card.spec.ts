import { mount } from "@vue/test-utils";

import Card from "@/components/Card.vue";
import PokemonType from "@/enums/pokemonType";

describe("Card.vue", () => {
  const build = (propsData?: object) => {
    const wrapper = mount(Card, {
      propsData: { buttonText: "Details", name: "Pokémon name", ...propsData },
      mocks: {
        $t: (text: string) => text
      }
    });

    const button = () => wrapper.find(".card__button");
    const image = () => wrapper.find(".card__image");
    const name = () => wrapper.find(".card__name");
    const number = () => wrapper.find(".card__number");
    const types = () => wrapper.findAll(".card__type");

    return { button, image, name, number, types, wrapper };
  };

  it("renders the component", () => {
    const { wrapper } = build();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("renders a card name", () => {
    const nameText = "Pikachu";
    const { name } = build({ name: nameText });

    expect(name().text()).toMatch(nameText);
  });

  it("renders a pokémon number", () => {
    const numberValue = 25;
    const { number } = build({ number: numberValue });

    expect(number().text()).toMatch(numberValue.toString());
  });

  it("renders pokémon types list", () => {
    const typesListValues = [PokemonType.Electric];

    const { types } = build({ types: typesListValues });

    expect(types().length).toEqual(typesListValues.length);
    expect(
      types()
        .at(0)
        .text()
    ).toMatch(PokemonType.Electric.toString());
  });

  it("renders a button text", () => {
    const buttonText = "Details";
    const { button } = build({ buttonText });

    expect(button().text()).toMatch(buttonText);
  });

  it("emits on-click event", async () => {
    const { button, wrapper } = build();

    await button().trigger("click");

    expect(wrapper.emitted("on-click")).toBeTruthy();
  });

  it("renders a card image", () => {
    const imageUrl = "https://images.pokemontcg.io/np/35.png";
    const { image } = build({ imageUrl });

    expect(image().exists()).toBe(true);
    expect(image().attributes("src")).toMatch(imageUrl);
  });
});
