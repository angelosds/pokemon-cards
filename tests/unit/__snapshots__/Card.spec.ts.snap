// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card.vue renders the component 1`] = `
<article class="card">
  <!---->
  <div class="card__description">
    <h1 class="card__name">Pok√©mon name</h1> <span class="card__number"></span>
    <!----> <button type="button" class="primary-button card__button">Details</button>
  </div>
</article>
`;
